File: 1/1-19.hs
-- Function to compute the transformation T_pq applied twice
fibMatrixMult :: (Int, Int) -> (Int, Int)
fibMatrixMult (p, q) = (p * p + q * q, 2 * p * q + q * q)

-- Function to compute the nth Fibonacci number in logarithmic time
fibFast :: Int -> Int
fibFast n = fibIter 1 0 0 1 n
  where
    fibIter a b p q count
      | count == 0 = b
      | even count = fibIter a b p' q' (count `div` 2)
      | otherwise  = fibIter (b * q + a * q + a * p) (b * p + a * q) p q (count - 1)
      where
        (p', q') = fibMatrixMult (p, q)

main :: IO ()
main = do
  putStrLn "Enter the value of n: "
  n <- readLn
  putStrLn $ "The " ++ show n ++ "th Fibonacci number is: " ++ show (fibFast n)

File: 1/1-19.cpp
#include <iostream>
#include <utility>

// Function to compute the transformation T_pq applied twice
std::pair<int, int> fibMatrixMult(int p, int q) {
    int pPrime = p * p + q * q;
    int qPrime = 2 * p * q + q * q;
    return std::make_pair(pPrime, qPrime);
}

// Function to compute the nth Fibonacci number in logarithmic time
int fibFast(int n) {
    // Helper function to perform the iterative process
    auto fibIter = [](int a, int b, int p, int q, int count) {
        if (count == 0)
            return b;
        else if (count % 2 == 0) {
            auto [pPrime, qPrime] = fibMatrixMult(p, q);
            return fibIter(a, b, pPrime, qPrime, count / 2);
        } else {
            return fibIter(b * q + a * q + a * p, b * p + a * q, p, q, count - 1);
        }
    };

    return fibIter(1, 0, 0, 1, n);
}

int main() {
    int n;
    std::cout << "Enter the value of n: ";
    std::cin >> n;
    std::cout << "The " << n << "th Fibonacci number is: " << fibFast(n) << std::endl;
    return 0;
}